python/lsst/ip/diffim/subtractImages.py

```diff
@@ -617,6 +605,232 @@ def _sourceSelector(self, sources):
 
         return selectSources.copy(deep=True)
 
+    def prepareInputs(self, template, science, sources,
+                      forceCompatibility=False,
+                      finalizedPsfApCorrCatalog=None):
+        self._validateExposures(template, science)
+        if self.config.doApplyFinalizedPsf:
+            self._applyExternalCalibrations(science,
+                                            finalizedPsfApCorrCatalog=finalizedPsfApCorrCatalog)
+        checkTemplateIsSufficient(template, self.log,
+                                  requiredTemplateFraction=self.config.requiredTemplateFraction)
+
+        # put the template on the same photometric scale as the science image
+        photoCalib = template.getPhotoCalib()
+        self.log.info("Applying photometric calibration to template: %f", photoCalib.getCalibrationMean())
+        template.maskedImage = photoCalib.calibrateImage(template.maskedImage)
+
+        if self.config.doScaleVariance and not forceCompatibility:
+            # Scale the variance of the template and science images before
+            # convolution, subtraction, or decorrelation so that they have the
+            # correct ratio.
+            templateVarFactor = self.scaleVariance.run(template.maskedImage)
+            sciVarFactor = self.scaleVariance.run(science.maskedImage)
+            self.log.info("Template variance scaling factor: %.2f", templateVarFactor)
+            self.metadata.add("scaleTemplateVarianceFactor", templateVarFactor)
+            self.log.info("Science variance scaling factor: %.2f", sciVarFactor)
+            self.metadata.add("scaleScienceVarianceFactor", sciVarFactor)
+
+        selectSources = self._sourceSelector(sources)
+        self.log.info("%i sources used out of %i from the input catalog", len(selectSources), len(sources))
+        if len(selectSources) < self.config.makeKernel.nStarPerCell:
+            self.log.warning("Too few sources to calculate the PSF matching kernel: "
+                             "%i selected but %i needed for the calculation.",
+                             len(selectSources), self.config.makeKernel.nStarPerCell)
+            raise RuntimeError("Cannot compute PSF matching kernel: too few sources selected.")
+
+        return selectSources
+
+
+class AlardLuptonPreconvolveSubtractConnections(SubtractInputConnections,
+                                                SubtractScoreOutputConnections):
+    pass
+
+    def __init__(self, *, config=None):
+        super().__init__(config=config)
+        if not config.doApplyFinalizedPsf:
+            self.inputs.remove("finalizedPsfApCorrCatalog")
+
+
+class AlardLuptonPreconvolveSubtractConfig(lsst.pipe.base.PipelineTaskConfig,
+                                           pipelineConnections=AlardLuptonPreconvolveSubtractConnections):
```

Comment:
The imports of the base config class become a little more complicated this way because of the connections, but maybe that is a worthwhile trade-off in order to reduce config duplication.